ksqldb-engine/src/main/java/io/confluent/ksql/services/DefaultConnectClientFactory.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.services;
+
+import io.confluent.ksql.services.ConnectClient.ConnectClientFactory;
+import io.confluent.ksql.util.FileWatcher;
+import io.confluent.ksql.util.KsqlConfig;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Paths;
+import java.util.Base64;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Properties;
+import org.apache.kafka.common.config.ConfigException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultConnectClientFactory implements ConnectClientFactory {
+
+  private static final Logger log = LoggerFactory.getLogger(DefaultConnectClientFactory.class);
+
+  private final KsqlConfig ksqlConfig;
+  private Optional<String> connectAuthHeader;
+  private FileWatcher credentialsFileWatcher;
+
+  public DefaultConnectClientFactory(
+      final KsqlConfig ksqlConfig
+  ) {
+    this.ksqlConfig = Objects.requireNonNull(ksqlConfig, "ksqlConfig");
+  }
+
+  @Override
+  public DefaultConnectClient get(final Optional<String> ksqlAuthHeader) {
+    if (connectAuthHeader == null) {
```

Comment:
I would probably modify this class such that the header and the watcher are created in the constructor, and we get rid of this `if`.  Right now, this class is not thread-safe in a way that allows starting multiple watchers, then losing references to all the old watchers, probably other phenomena as well.

Reply:
We can't create the header and watcher in the constructor -- the DefaultConnectClientFactory constructor is called every time a service context is created, i.e., on every ksql request, regardless of whether it's a connector request or not. We don't want to load credentials (and potentially start the watcher) on every request; we only want to do this if a connector request is issued. 

I can synchronize this `get()` method in order to address the concurrency concern you raised -- `get()` should only ever be called from a single thread anyway.

Reply:
Makes sense! FWIW we should also synchronize `close`.